File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] nums, int target) {
        List<List<Integer>> res= new ArrayList<>();
        Arrays.sort(nums);
        dfs(nums,new ArrayList<>(), target, res);
        return res;
    }
    
    public void dfs(int[] nums, List<Integer> path, int target, List<List<Integer>> res ){
        if(target==0)
            res.add(path);
        for (int i=0; i<nums.length;i++){
            if (i>0 && nums[i]==nums[i-1]){
                continue;
            }
            if (nums[i]>target)
                break;
            List<Integer> c= new ArrayList<>(path);
            c.add(nums[i]);
            dfs(Arrays.copyOfRange(nums,i+1,nums.length), c, target-nums[i],res);
        }
    }
}

File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, nums: List[int], target: int) -> List[List[int]]:
        res=[]
        nums.sort()
        self.dfs(nums,target,[],res)
        return res
        
    def dfs(self,nums,target,path,res):
        if target==0:
            res.append(path)
            return

        for i in range(len(nums)):
            if i>0 and nums[i-1]==nums[i]:
                continue
            if nums[i]>target:
                break
            self.dfs(nums[i+1:],target-nums[i],path+[nums[i]],res)